app/com/arpnetworking/rollups/RollupGenerator.java

```diff
@@ -232,53 +250,73 @@ private void handleLastDataPointMessage(final LastDataPointMessage message) {
                     ActorRef.noSender());
         } else {
             _metrics.recordCounter("rollup/generator/last_data_point_message/success", 1);
-            final Instant startOfRecentClosedPeriod = period.recentStartTime(_clock.instant());
-            final Instant lastDataPoint = message.getLastDataPointTime().orElse(Instant.EPOCH);
+
+            // Example:
+            //
+            // Consider a minutely metric that has just hit 00:00 UTC 3 Jan
+            //
+            //                 22:00         23:00         00:00      startOfLastEligiblePeriod
+            //                   |             |             |
+            // minutely  x x x x x x x x x x x x x x x x x x x                  N/A
+            //   hourly          x             x             |        23:00 2 Jan (1 period  ago)
+            //    daily          |             |             |        00:00 1 Jan (2 periods ago)
+            //
+            // Hourly pulls from the minutely, and sees that the most recently closed period
+            // is before the most recent datapoint in minutely, and so we can roll up.
+            //
+            // Daily pulls from hourly, but the latest hourly datapoint is at 23:00 < 00:00, the end
+            // of the most recent period. Therefore we cannot roll-up just yet.
+
+            final Instant lastRollupDataPoint = message.getRollupLastDataPointTime().orElse(Instant.EPOCH);
+
+            final Instant startOfLastEligiblePeriod =
+                message.getSourceLastDataPointTime()
+                        .map(period::recentStartTime)
+                        .orElse(Instant.EPOCH);
 
             // If the most recent period aligned start time is after the most recent datapoint then
             // we need to run the rollup, otherwise we can skip this and just send a finish message.
-            if (startOfRecentClosedPeriod.isAfter(lastDataPoint)) {
-                final String rollupMetricName = metricName + period.getSuffix();
+            if (startOfLastEligiblePeriod.isAfter(lastRollupDataPoint)) {
                 final int maxBackFillPeriods = _maxBackFillByPeriod.getOrDefault(period, 0);
 
                 final Instant oldestBackfillPoint = period.recentEndTime(_clock.instant())
                         .minus(period.periodCountToDuration(maxBackFillPeriods));
 
                 // We either want to start at the oldest backfill point or the start of the period
                 // after the last datapoint since it contains data for the period that follows it.
-                Instant rollupPeriodStart = lastDataPoint.isBefore(oldestBackfillPoint)
-                        ? oldestBackfillPoint : period.recentEndTime(lastDataPoint).plus(period.periodCountToDuration(1));
+                Instant rollupPeriodStart = lastRollupDataPoint.isBefore(oldestBackfillPoint)
+                        ? oldestBackfillPoint : period.recentEndTime(lastRollupDataPoint).plus(period.periodCountToDuration(1));
```

Comment:
> `period.recentEndTime(lastRollupDataPoint)`

Why is the `period.recentEndTime()` necessary? It [looks to me](https://github.com/ArpNetworking/metrics-portal/blob/master/app/com/arpnetworking/rollups/RollupPeriod.java#L50) like `recentEndTime` just aligns a time, but... `lastRollupDataPoint` should already _be_ aligned, right?

Reply:
That code is defensive, but you're correct – if it's coming from the rollup it should already be aligned.